src/Compilers/Core/Portable/InternalUtilities/OneOrMany.cs

```diff
@@ -35,6 +36,43 @@ public OneOrMany(ImmutableArray<T> many)
             _many = many;
         }
 
+        /// <summary>
+        /// Maps each item in this OneOrMany using the map function.
+        /// </summary>
+        public OneOrMany<TResult> SelectAsOneOrMany<TResult>(Func<T, TResult> map)
+        {
+            if (_many.IsDefault)
+            {
+                return OneOrMany.Create(map(_one));
+            }
+            else
+            {
+                return OneOrMany.Create(_many.SelectAsArray(map));
+            }
+        }
+
+        /// <summary>
+        /// Maps each item in this OneOrMany using the map function and the corresponding argument from args.
+        /// </summary>
+        internal OneOrMany<TResult> SelectAsOneOrMany<TArgs, TResult>(Func<T, TArgs, TResult> map, OneOrMany<TArgs> args)
```

Comment:
The name was by imitation with `ImmutableArray.SelectAsArray` (the one that takes args). But I agree. I'll rename to `ZipAsOneOrMany`.